src/nbl/video/CVulkanLogicalDevice.h

```diff
@@ -304,92 +324,249 @@ class CVulkanLogicalDevice final : public ILogicalDevice
+        }
+        createInfo.pDependencies = deps.data();
+
+        // vk->vk.vkCreateRenderPass(vkdev, &ci, nullptr, &m_renderpass);
+        VkRenderPass vk_renderpass;
+        if (vkCreateRenderPass(m_vkdev, &createInfo, nullptr, &vk_renderpass) == VK_SUCCESS)
+        {
+            return core::make_smart_refctd_ptr<CVulkanRenderpass>(
+                core::smart_refctd_ptr<CVulkanLogicalDevice>(this), params, vk_renderpass);
+        }
+        else
+        {
+            return nullptr;
+        }
     }
-            
+           
+    // API needs to change, vkFlushMappedMemoryRanges could fail.
     void flushMappedMemoryRanges(core::SRange<const video::IDriverMemoryAllocation::MappedMemoryRange> ranges) override
     {
-        return;
+        constexpr uint32_t MAX_MEMORY_RANGE_COUNT = 408u;
+        VkMappedMemoryRange vk_memoryRanges[MAX_MEMORY_RANGE_COUNT];
+
+        const uint32_t memoryRangeCount = static_cast<uint32_t>(ranges.size());
+        assert(memoryRangeCount <= MAX_MEMORY_RANGE_COUNT);
+
+        getVkMappedMemoryRanges(vk_memoryRanges, ranges.begin(), ranges.end());
+        
+        if (vkFlushMappedMemoryRanges(m_vkdev, memoryRangeCount, vk_memoryRanges) != VK_SUCCESS)
+            printf("flushMappedMemoryRanges failed\n");
     }
             
+    // API needs to change, this could fail
     void invalidateMappedMemoryRanges(core::SRange<const video::IDriverMemoryAllocation::MappedMemoryRange> ranges) override
     {
-        return;
+        constexpr uint32_t MAX_MEMORY_RANGE_COUNT = 408u;
+        VkMappedMemoryRange vk_memoryRanges[MAX_MEMORY_RANGE_COUNT];
+
+        const uint32_t memoryRangeCount = static_cast<uint32_t>(ranges.size());
+        assert(memoryRangeCount <= MAX_MEMORY_RANGE_COUNT);
+
+        getVkMappedMemoryRanges(vk_memoryRanges, ranges.begin(), ranges.end());
+
+        if (vkInvalidateMappedMemoryRanges(m_vkdev, memoryRangeCount, vk_memoryRanges) != VK_SUCCESS)
+            printf("invalidateMappedMemoryRanges failed!\n");
     }
 
     bool bindBufferMemory(uint32_t bindInfoCount, const SBindBufferMemoryInfo* pBindInfos) override
     {
+        bool anyFailed = false;
         for (uint32_t i = 0u; i < bindInfoCount; ++i)
         {
-            if ((pBindInfos[i].buffer->getAPIType() != EAT_VULKAN) /*|| (pBindInfos[i].memory->getAPIType() != EAT_VULKAN)*/)
+            const auto& bindInfo = pBindInfos[i];
+            
+            if ((bindInfo.buffer->getAPIType() != EAT_VULKAN) || (bindInfo.memory->getAPIType() != EAT_VULKAN))
                 continue;
 
-            VkBuffer vk_buffer = static_cast<const CVulkanBuffer*>(pBindInfos[i].buffer)->getInternalObject();
+            CVulkanBuffer* vulkanBuffer = static_cast<CVulkanBuffer*>(bindInfo.buffer);
+            vulkanBuffer->setMemoryAndOffset(
+                core::smart_refctd_ptr<IDriverMemoryAllocation>(bindInfo.memory), bindInfo.offset);
```

Comment:
remember that its also an error to bind memory more than once to a buffer (unless the buffer is sparse, but then its a sparse bind request which is something completely different)